[WIP] link: sync deps with parent package

schomatis opened the pull request:
This PR extracts the `--sync` feature from https://github.com/whyrusleeping/gx-go/pull/48 without the `vendor/` support (that has some implementation problems, see https://github.com/whyrusleeping/gx-go/issues/46#issuecomment-416229688).

As the imported PR, this command now requires `gx-go link` to be run inside the parent package of the dependency even though the link is still done to the global workspace. This breaks API but it makes easier to reason about the code and about the link being done, and in most use cases the developer was already calling the command from a package.

Closes #46.

schomatis commented:
(Another possible enhancement for this PR or a new one: detect already linked packages and un-link+link them again. At the moment it would seem that the `post-install` hook + rewrite functionality doesn't re-process an already rewritten file -it's assuming that the import paths wont change?, but if we've updated something?- so if I link without `--sync` and it fails due to a mismatch version, when I link again, now with `--sync`, the rewrite part of the process won't happen but it won't fail either. The result is that both with and without the sync functionality will seem to fail -the user will need to actually un-link and re-link manually, but this won't be obvious to him/her-.)

Stebalien commented:
> detect already linked packages and un-link+link them again.

We *should* probably do that but we can do it in a new patch.

schomatis commented on link.go:
```diff
@@ -19,30 +19,34 @@ var LinkCommand = cli.Command{
 
-Example workflow:
+Links the specified dependency package (by name or hash) used by the
+parent package (in the CWD or specified by --parent). The link is done
```
I'd prefer to maintain the parent-child relationship concept, because everything is a package here  *but*, depending on where you're looking from (in what part of the code you're in), that package is either a dependency (and that distinction is important because a dependency in a `package.json` doesn't have the same information as the actual package) or the package that's importing that dependency. The term "current" doesn't seem to carry that relationship meaning, is there another term you'd prefer?

Stebalien replied:
When I'm using GX, I'm usually working *on* a gx package. The "current" package is the package I'm currently working on.

For example:

> `gx-go link` replaces the target gx package (either by name or hash) with a symlink to the appropriate dvcs repo in your GOPATH. To make this "symlinked" repo usable as a gx package, `gx-go link` rewrites the target package's dvcs imports using the target package's `package.json`. Unfortunately, this can cause build errors in packages *depending* on this package if these dependent packages specify alternative, conflicting dependency versions. We can work around this by asking `gx` to rewrite the target package using dependencies from the current package (the package you're trying to build) first, falling back on the target package's `package.json` file. We can do this by passing the `--deps=mine` flag.

That *is* a bit wordy and you're right, "current" may not be quite right.

schomatis replied:
Let's leave it at "the current dependent package". Nice text (I like it *wordy* :), I'll add it to the usage doc.

schomatis commented on link.go:
```diff
@@ -53,21 +57,33 @@ unlinked QmVGtdTZdTFaLsaj2RwdVG8jcjNNcp1DE914DKZ2kHmXHw /home/user/go/src/github
+			Usage: "Specify the path of the parent package of the linked dependency (default: CWD).",
+		},
+		cli.BoolFlag{
+			Name:  "s,sync",
```
Yes, I don't really like "sync" but couldn't come up with something better. Let's go with `--deps=` and use the term that will replace "parent" here, e.g., `--deps=<package-that-is-including-this-package>`.

schomatis commented on link.go:
```diff
@@ -53,21 +57,33 @@ unlinked QmVGtdTZdTFaLsaj2RwdVG8jcjNNcp1DE914DKZ2kHmXHw /home/user/go/src/github
+		},
+		cli.BoolFlag{
+			Name:  "s,sync",
+			Usage: "Synchronize the dependencies of the linked package with the ones of its parent.",
```
Would you prefer to use "dependent"?

Stebalien replied:
I would, but then you have to answer the question "which dependent". The answer is "the current one".

Stebalien commented:
Thinking about this, we could actually use the dependencies specified in the *original* gxed package (the one we're replacing with a symlink). That would be the most conservative approach (effectively "downgrading" dependencies.

The only difference is that, if the target repo has added a new dependency *also* present in the current/parent package, we'll use the version in the *target* repo, not the current/parent. However, using the package.json of the current/parent isn't foolproof either (e.g., the target could have added a dependency that we *don't* have, one that depends on packages we *do* depend on but depends on different versions).


The other approach would be to do a fake gx-workspace update. However, that's *way* more work.

schomatis commented:
> The only difference is that, if the target repo has added a new dependency also present in the current/parent package, we'll use the version in the target repo, not the current/parent. However, using the package.json of the current/parent isn't foolproof either (e.g., the target could have added a dependency that we don't have, one that depends on packages we do depend on but depends on different versions).

Let's have both options: `--deps=dependent` and `--deps=dependency` (we can use other terms but let's keep them absolute, not relative to the current/mine/theirs perspective) and recommend the second one (and have it as the default) as the conservative approach (noting that it not foolproof).

Stebalien commented:
> Let's have both options: --deps=dependent and --deps=dependency (we can use other terms but let's keep them absolute, not relative to the current/mine/theirs perspective) and recommend the second one (and have it as the default) as the conservative approach (noting that it not foolproof).

In this case, the other strategy would really be "original" or "replaced".

I'm really confused why you feel so strongly about keeping the terminology absolute. Personally, I find it *extremely* confusing (depending on the context). When I see an absolute term, I think "oh, they must *not* mean my current package, otherwise they would have just said that". Specifically, when I see "dependent", my first thought is "which dependent?".

schomatis commented:
> I'm really confused why you feel so strongly about keeping the terminology absolute. Personally, I find it extremely confusing (depending on the context).

Ok, let me rethink that and get back to you.

I think that my main objection is about the code and not the UX. From the point of the user who just runs `cd package; gx-go link dependency` I think you're right, "current" is the package we just moved to, there's no doubt there. Using "current" in the code starts losing meaning as we're constantly changing perspectives between dependency-dependent packages, and the original ("current") package from where we (the user) started has little to do with the ongoing operation of, for example, fetching the dependency of the the dependency. That's where I want to have a clear model of which package is on top from where we are extracting its dependencies to later convert (fetch) them into other (also) packages. But yes, that part of the code is a different realm from the command the user runs.

Stebalien commented:
Ah, yes, *that's* totally reasonable. `parent` actually seems like a good word in that case (or `root`).

schomatis commented:
@Stebalien Could you take another look please? Changed the terminology to target/current (using your command description) and renamed the option `--override-deps`. I've finally had to settle for just using the versions from the current dependent package (and didn't add the variation of taking them from the target package because this command overwrites it -since everything is done in the same global directory- and wanted to avoid changing too much functionality in this first iteration).

schomatis commented:
@Stebalien

schomatis commented:
@Stebalien
